Src/GoogleApis.Tests/Apis/Requests/PageStreamerTest.cs

```diff
@@ -0,0 +1,241 @@
+            new Page(null, 4, 5))
+        {
+            RequestCheck = "foo"
+        };
+        
+        private static PageStreamedResource[] AllResources = { simpleResource, resourceWithEmptyPages, requestCheckingResource };
+
+        [Test]
+        [TestCaseSource("AllResources")]
+        public void Fetch(PageStreamedResource resource)
+        {
+            var actual = PageStreamedResource.PageStreamer.Fetch(resource.NewRequest());
+            Assert.AreEqual(resource.AllItems, actual);
+        }
+
+        [Test]
+        [TestCaseSource("AllResources")]
+        public async Task FetchAllAsync(PageStreamedResource resource)
+        {
+            var actual = await PageStreamedResource.PageStreamer.FetchAllAsync(
+                resource.NewRequest(),
+                CancellationToken.None);
+            Assert.AreEqual(resource.AllItems, actual);
+        }
+
+        // Note: this doesn't test all the possible timings of cancellation
+        [Test]
+        public async Task Cancellation()
+        {
+            var cts = new CancellationTokenSource();
+            var gatekeeperSource = new TaskCompletionSource<int>();
+            var resource = new PageStreamedResource(
+                "check",
+                new Page("x", 1, 2, 3),
+                new Page("y", 4, 5),
+                new Page(null, 6, 7));
+            resource.gateKeeper = gatekeeperSource.Task;            
+
+            var task = PageStreamedResource.PageStreamer.FetchAllAsync(resource.NewRequest(), cts.Token);
+            // Delay the resource fetching until we've cancelled the token
+            cts.Cancel();
+            gatekeeperSource.SetResult(0);
+            // TODO: Move to Assert.ThrowsAsync when we update to an appropriate version of NUnit.
+            // Assert.That is available, but the documentation is unclear how this interacts with await.
+            try
+            {
+                await task;
+                Assert.Fail("Expected exception");
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected
+            }
+        }
+
+        // Has to be public so we can use it as a parameter for test cases
+        public class PageStreamedResource
+        {
+            internal static readonly PageStreamer<int, Request, Page, string> PageStreamer =
+                new PageStreamer<int, Request, Page, string>(
+                    (request, token) => request.WithToken(token),
+                    page => page.NextPageToken,
+                    page => page.Items);
+
+            // Task to await when fetching pages; defaults to an already completed task,
+            // but can be set to another task to allow for more fine-grained interaction.
+            internal Task gateKeeper = TaskEx.FromResult(0);
+            internal List<Page> Pages { get; private set; }
+            internal string Name { get; set; }
+            internal string RequestCheck { get; set; }
```

Comment:
It's used to test that the request we actually get is the one we expected. It's possibly not worth the extra test complexity though...